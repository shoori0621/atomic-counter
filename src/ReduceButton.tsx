import React from 'react';
import {Text, StyleSheet, TouchableOpacity} from 'react-native';

interface Props {
  count: number;
  setCounter: (count: number) => void;
}

export default function ReduceButton(props: Props) {
  const {count, setCounter} = props;
  const reduceCounter = React.useCallback(() => {
    setCounter(count - 1);
  }, [count, setCounter]);
  return (
    <TouchableOpacity
      style={[styles.button, count <= 0 && styles.disabledButton]}
      onPress={reduceCounter}
      disabled={count <= 0}>
      <Text style={styles.buttonText}>-</Text>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#008080',
    width: 50,
    height: 50,
    justifyContent: 'center',
    alignItems: 'center',
  },
  buttonText: {
    color: 'white',
    fontSize: 20,
  },
  disabledButton: {
    backgroundColor: 'gray',
  },
});
