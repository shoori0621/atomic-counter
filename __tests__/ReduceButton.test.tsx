import React from 'react';
import renderer from 'react-test-renderer';
import {fireEvent, render, screen} from '@testing-library/react-native';
import ReduceButton from '../src/ReduceButton';

// Note: import explicitly to use the types shiped with jest.
import {it, describe, expect, jest} from '@jest/globals';

describe('ReduceButton', () => {
  describe('countが0より大きい時', () => {
    const count = 10;
    it('通常のReduceButtonを描画できる', () => {
      const component = renderer.create(
        <ReduceButton count={count} setCounter={() => {}} />,
      );
      expect(component).toMatchSnapshot();
    });
    it(`ボタン押下でカウントが${count}から${count - 1}になる`, () => {
      const onPressEvent = jest.fn();
      render(<ReduceButton count={count} setCounter={onPressEvent} />);
      fireEvent.press(screen.getByTestId('reduce-button'));
      expect(onPressEvent).toBeCalled();
      expect(onPressEvent.mock.calls[0][0]).toBe(count - 1);
    });
  });
  describe('countが0の時', () => {
    const count = 0;
    it('グレーアウトされたReduceButtonを描画できる', () => {
      const component = renderer.create(
        <ReduceButton count={count} setCounter={() => {}} />,
      );
      expect(component).toMatchSnapshot();
    });
    it(`ボタン押下できない`, () => {
      const onPressEvent = jest.fn();
      render(<ReduceButton count={count} setCounter={onPressEvent} />);
      fireEvent.press(screen.getByTestId('reduce-button'));
      expect(onPressEvent).not.toBeCalled();
    });
  });
});
