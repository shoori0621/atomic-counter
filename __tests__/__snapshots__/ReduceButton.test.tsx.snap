// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduceButton countが0の時 グレーアウトされたReduceButtonを描画できる 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "gray",
      "height": 50,
      "justifyContent": "center",
      "opacity": 1,
      "width": 50,
    }
  }
  testID="reduce-button"
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 20,
      }
    }
  >
    -
  </Text>
</View>
`;

exports[`ReduceButton countが0より大きい時 通常のReduceButtonを描画できる 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#008080",
      "height": 50,
      "justifyContent": "center",
      "opacity": 1,
      "width": 50,
    }
  }
  testID="reduce-button"
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 20,
      }
    }
  >
    -
  </Text>
</View>
`;
